### DISCUSSION TITLE
Data loss from multiple calls to CompleteMultipartUpload object in a short period of time

### DISCUSSION THREAD

**Discussion by xieys0** [minio/minio#17859] [Q&A]
## Environment
OS:  Ubuntu22.04.1
Version:  https://dl.min.io/server/minio/release/linux-amd64/archive/minio_20230809233022.0.0_amd64.deb

# df -HT
root@ubuntu:~# df -hT
Filesystem     Type   Size  Used Avail Use% Mounted on
tmpfs          tmpfs  6.3G  2.2M  6.3G   1% /run
/dev/sda2      ext4   479G   12G  443G   3% /
tmpfs          tmpfs   32G     0   32G   0% /dev/shm
tmpfs          tmpfs  5.0M     0  5.0M   0% /run/lock
/dev/sda1      vfat   1.1G  6.1M  1.1G   1% /boot/efi
tmpfs          tmpfs  6.3G  4.0K  6.3G   1% /run/user/0
/dev/sdb       xfs    7.3T   52G  7.3T   1% /mnt/disk1
/dev/sdc       xfs    7.3T   52G  7.3T   1% /mnt/disk2
/dev/sdd       xfs    7.3T   52G  7.3T   1% /mnt/disk3
/dev/sde       xfs    7.3T   52G  7.3T   1% /mnt/disk4

# cat /etc/default/minio
MINIO_ROOT_USER=xxxx
MINIO_ROOT_PASSWORD=xxxx
MINIO_OPTS="--address :9000 --console-address :9001"
MINIO_VOLUMES="/mnt/disk{1...4}"

## Describe
Calling the CompleteMultipartUpload interface multiple times for the same uploadId in a short period of time. The first call returned success, while the other returned uploadId does not exist. The object information can be found, but the object data is missing and cannot be downloaded or restored. There are only xl. meta files in the object folder.

root@ubuntu:/mnt/disk1/test/0CA985D3589645E49ADFAF4391FA754D# ls -la
total 4
drwxr-xr-x 2 minio-user minio-user  29 Aug 15 11:14 .
drwxr-xr-x 3 minio-user minio-user  54 Aug 15 11:14 ..
-rw-r--r-- 1 minio-user minio-user 844 Aug 15 11:14 xl.meta

## Expected Behavior
Object data is not lost

## Current Behavior
Object data is not lost

## Notes
Historical versions and Ubuntu multi node and multi driver environments can still be replicated

**Comment by harshavardhana** [comment]
> Calling the CompleteMultipartUpload interface multiple times for the same uploadId in a short period of time. The first call returned success, while the other returned uploadId does not exist. The object information can be found, but the object data is missing and cannot be downloaded or restored. There are only xl. meta files in the object folder.

Please provide a reproducer here @xieys0 until then this issue is nonactionable for us.

**Reply by xieys0** [reply]
s3_test.go

```go
import (
	"bytes"
	"context"
	"crypto/tls"
	"fmt"
	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"net"
	"net/http"
	"testing"
	"time"
)

// github.com/minio/minio-go/v7 v7.0.38
func TestDemo(t *testing.T) {

	// new client
	endpoint := "192.168.1.170:9000"
	region := "us-east-1"
	useSSL := false
	accessKey := "minioadmin"
	accessSecret := "minioadmin"
	minioClient, _ := minio.NewCore(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, accessSecret, ""),
		Secure: useSSL,
		Region: region,
		Transport: &http.Transport{
			Proxy: http.ProxyFromEnvironment,
			DialContext: (&net.Dialer{
				Timeout:   30 * time.Second,
				KeepAlive: 30 * time.Second,
			}).DialContext,
			MaxIdleConns:          256,
			MaxIdleConnsPerHost:   16,
			ResponseHeaderTimeout: time.Minute,
			IdleConnTimeout:       time.Minute,
			TLSHandshakeTimeout:   10 * time.Second,
			ExpectContinueTimeout: 10 * time.Second,
			DisableCompression:    true,
			TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
		},
	})

	// MultipartUpload Object
	bucket := "test" // Versioning OFF, Object Locking OFF, Quota OFF
	object := "testObject.data"
	uploadId, err := minioClient.NewMultipartUpload(context.Background(), bucket, object, minio.PutObjectOptions{})
	if err != nil {
		t.Error(err)
		return
	}
	fmt.Println("NewMultipartUpload end, uploadId=", uploadId) //NWM1MTZhMWYtZTNlMy00NjFiLTg1MjctNDE5ODM4ZDY2NGIxLmM4ZDdlNTdlLThjMDEtNGRiNC1iNmZhLWUyNzkwODhkM2JmYQ

	partSize := int64(1024 * 1024 * 5)
	partNums := 100
	parts := make([]minio.CompletePart, partNums, partNums)
	data := []byte{}
	for j := int64(0); j < partSize; j++ {
		data = append(data, 1)
	}
	for i := 0; i < partNums; i++ {
		partID := i + 1

		partData := make([]byte, len(data))
		copy(partData, data)

		reader := bytes.NewReader(partData)
		objPart, err := S3Core.PutObjectPart(context.Background(), bucket, object, uploadId, partID, reader, partSize, "", "", nil)
		if err != nil {
			t.Error(err)
		}
		parts[i] = minio.CompletePart{
			PartNumber: objPart.PartNumber,
			ETag:       objPart.ETag,
		}
	}
	fmt.Println("PutObjectPart end")

	res, err := S3Core.ListMultipartUploads(context.Background(), bucket, object, "", "", "", 1000)
	if err != nil {
		t.Error(err)
	}
	uploadId2 := ""
	if len(res.Uploads) > 0 {
		uploadId2 = res.Uploads[0].UploadID
	}
	fmt.Println("uploadId2=", uploadId2) //c8d7e57e-8c01-4db4-b6fa-e279088d3bfa

	ch1 := make(chan int)
	ch2 := make(chan int)
	go func() {
		etag, err := S3Core.CompleteMultipartUpload(context.Background(), bucket, object, uploadId, parts, minio.PutObjectOptions{})
		fmt.Println(etag, "  ", err) //6a7a32f75cec39d7f42e0d826dc636b7-100    <nil>
		ch1 <- 1
	}()
	go func() {
		etag, err := S3Core.CompleteMultipartUpload(context.Background(), bucket, object, uploadId2, parts, minio.PutObjectOptions{})
		fmt.Println(etag, "  ", err) //The specified key does not exist.
		ch2 <- 1
	}()
	<-ch1
	<-ch2
	fmt.Println("CompleteMultipartUpload end")
	// result getObject
	err = S3Core.FGetObject(context.Background(), bucket, object, object, minio.GetObjectOptions{
		ServerSideEncryption: nil,
		VersionID:            "",
		PartNumber:           0,
		Checksum:             false,
		Internal:             minio.AdvancedGetOptions{},
	})
	if err != nil {
		t.Error(err)//Resource requested is unreadable, please reduce your request rate
	}
}
```

**Reply by klauspost** [reply]
@harshavardhana This could be unintended cleanup from the second call. Opened an issue to track: https://github.com/minio/minio/issues/17863

We must handle these calls. This can happen with a client retry unintentionally.
